### PR TITLE
fix(UniversalController): rename run() since super.run() has greater arity

### PR DESCRIPTION
**Pull request by sladiri**
…ferent arity than super.run().

When using the Router module, the payload becomes undefined, because FunctionTree.run() is expected
to take at least three parameters, while currently UniversalController.run() takes only two.

ISSUES CLOSED: #979

### CODE REVIEW

**Comment by christianalfoni** [comment]
Hm.... I am thinking here that maybe running "a signal" is the wrong approach. What I have noticed after actually using this is that you will grab all your state using other mechanisms to go an grab the state. So basically you just want to prepare some state and the shove it into the controller.

So maybe we should rather do:

```js
controller.setState({
  'app.user': user,
  'admin.list': adminList
})
```

Improves readability as well as there is no promise return. Does this fit your scenario as well?

**Comment by sladiri** [comment]
@christianalfoni I get the value of updateState, that would be a good API, but in my case I want to update the model by calling the signals I defined. I thought that I could replace side effects via the context for example.

I have not yet tried much, but do you suggest that I run a FunctionTree instance, not the signals from the module/controller? But how would I pass in the model to such a FunctionTree, so that it would become the state for the actions?

**Comment by christianalfoni** [comment]
@sladiri Hm, yeah, allowing running signals is a cool feature... just replacing the providers with server side providers. 

There is nothing wrong doing both here really. It depends on your approach. Either running custom backend logic to extract state or actually implement providers to run signals. But maybe then that should be the separation?

Lets write a spec here:

1. We implement `setState` as a way to just set some state and use the controller

2. We implement `runSequence` as a way to just run a sequence and return the promise (runOnServer) basically

You wanna play with it or I can have a go at it :-)

**Comment by sladiri** [comment]
@christianalfoni I would like to have a go, yes. :)

**Comment by Guria** [comment]
cc @christianalfoni

**Comment by christianalfoni** [comment]
Great stuff @sladiri :-)

Two small last things: `getModelPathArray`, could you move it to the `utils` instead? Not a very useful public method on the controller :-)

Second just document the two new `UniversalController` methods in the docs? I can do this if you are unsure how to add it :-)

**Comment by sladiri** [comment]
@christianalfoni Certainly, now I see again that `runSequence` is just a convenience delegator, which sets a default name for a signal. (I have not read the docs well enough.) Is this additional method really required? :) I guess everyone would just use `run`, and use their own name, wouldn't they?

File: packages/node_modules/cerebral/src/utils.js [hunk 1]
```diff
@@ -334,3 +334,7 @@ export function getStateTreeProp(props = {}) {
     return hit
   }, null)
 }
+
+export function getModelPathArray(path) {
```
Comment (yusufsafak) [inline packages/node_modules/cerebral/src/utils.js]: @sladiri code duplication :) you can just use utils.ensurePath  :)

**Comment by yusufsafak** [comment]
@sladiri you can use utils.ensurePath instead of creating getModelPathArray !

**Comment by christianalfoni** [comment]
@yusufsafak Oh, for sure, thanks for that :)

**Comment by sladiri** [comment]
Nice catch thank you. :)

**Comment by sladiri** [comment]
I guess that manually getting the signal out of the controller is not ideal? :)
Maybe it should accept a name or an array? The name would be used to get the signal inside `runSequence` with `Controller.getSignal`.

**Comment by sladiri** [comment]
In my last change I added the possibility to pass in a signal-path to runSequence. I hope it is OK, I find it cumbersome to extract a predefined module-signal manually from the controller object.

**Comment by sladiri** [comment]
Nice, that was no problem :) Thank you.

File: docs/api/universalcontroller.md [hunk 2]
```diff
@@ -43,10 +43,50 @@ controller.run([
...
+})
+
+controller
+  .runSequence('app.aSignal', {isAwesome: true})
+  .then(() => {
+    // I am done running
+  })
+```
+
 You can add any providers to the controller to do database fetching etc. inside this **run** execution. Think of it as a signal the updates the state of the app before rendering it on the server.
```
Comment (Guria) [inline docs/api/universalcontroller.md]: Should **run** be chaged to **runSequence** here?

Comment (sladiri) [inline docs/api/universalcontroller.md]: It seems so, yes. I missed that while copying the old text.

**Comment by christianalfoni** [comment]
Moved this PR to https://github.com/cerebral/cerebral/pull/1020, due to current security issues and also rebasing

**Pull request closed without merging**